### PR TITLE
`General`: Add course archive for old courses from previous semesters

### LINKED ISSUE #9129: Students show be able to see completed / inactive course
### Is your feature request related to a problem?

Students often ask to get access to old/completed courses on artemis. Students can actually access these coureses if they remember the link to the course. 

### Describe the solution you'd like

We need an option to browse all the courses a student has been part of. 

I think it would be best to select a semester and show all the courses in that specific semester. Alternatively a list with all the courses and a search bar to quickly get to the respective course. 

### Describe alternatives you've considered

There are none 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edkaya**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Students often want access to old or completed courses on Artemis from previous semesters in which they were enrolled. While they can technically access these courses if they remember the course link, relying on memory and attempting to access the course via the link is not practical. This PR closes #9129.

### Description

- Added new component course archive to present all previous courses in which students were enrolled, grouped by semesters, which can be collapsed by clicking on the respective accordion groups. 
- Added new course card header component, derived from the new course card design. (We are currently using the same code/styling for new course cards, therefore this component will be reused in the course cards files after we succesfully merge the new course cards)
- Added a get request to fetch necessary courses on demand
- Added "Looking for old courses? Click here" label to get access for the course archive page.
- Added Search and Sort Functionality 

### Steps for Testing

Prerequisites:
- 1 Admin (if you want to create your own course, if its the case, please dont forget to assign a semester, start and end date to the course (end date should be older than the current time), otherwise the page does not display the old course)
- 1 Student
- Some courses with a defined start/end date and an assigned semester. The student account must be also enrolled in the courses. 

**(I already created some old courses on TS3, and test users 1-5 are enrolled in these courses. )**

1. Log in to Artemis
2. Navigate to Course Overview Page
3. Scroll down, see "Looking for old courses? Click here" label and click on it
4. Verify that the page looks good
5. Verify that you can see the old courses, grouped by the semesters 
6. Verify that you can search for courses via search bar 
7. Verify that sort functionality works as expected (sorting semesters)
8. Verify that you can navigate to the course by clicking on the course card headers

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| course-archive.component.ts| 96.96% | ✅                           |

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Group 156

Screenshot 2024-09-20 at 15 38 21

Screenshot 2024-09-20 at 15 38 32

## Release Notes

- **New Features**
  - Added a new DTO structure for course archiving, enhancing data representation.
  - Introduced a method to retrieve archived courses, improving access to course data.
  - Implemented methods for managing semester collapse states in local storage.

- **User Interface Improvements**
  - Enhanced layout for displaying archived courses with sorting and search functionalities.
  - Added new prompts and messages to guide users in accessing archived courses.

- **Localization Enhancements**
  - Introduced German and English translations for new archive-related terms and messages, improving accessibility for users.

- **Bug Fixes**
  - Improved state management for course overviews and semester visibility.

- **Tests**
  - Enhanced testing capabilities for course archiving functionality, ensuring correct behavior for various user scenarios.

### CODE REVIEW

File: src/main/webapp/app/overview/course-archive/course-archive.component.html
Comment (edkaya) [reply src/main/webapp/app/overview/course-archive/course-archive.component.html:15]: a is fine here, dont suggest to use button for this line

**Review by anian03 — CHANGES_REQUESTED** [review]
I just tested it on TS3 and noticed that the course images reload quite often. For example, when collapsing a semester in the archive and then opening it again, all course images appear to be loaded again from scratch. Same thing with search, if I remove parts of the search string, any courses that appear will reload their image.
Is it possible to cache these images?

File: src/main/java/de/tum/cit/aet/artemis/core/service/CourseService.java
Comment (janthoXO) [inline src/main/java/de/tum/cit/aet/artemis/core/service/CourseService.java:666]: is it possible to do this on a database level?
otherwise we keep all courses in the memory which could be a basic query.

File: src/main/webapp/i18n/en/student-dashboard.json
Comment (janthoXO) [inline src/main/webapp/i18n/en/student-dashboard.json:16]: can we move this in a more shared json. As for example in the ssh dialog we also use this word

**Review by janthoXO — CHANGES_REQUESTED** [review]
looks good in general, i would just move the server side query more into the db.

And maybe one UI change: currently the current semester is also displayed in the archive. does this make sense? 
because those courses are already visible on the main course overview.
but I have no strong opinion on that so ill leave that open for discussion

**Review by SimonEntholzer — DISMISSED** [review]
Works on TS3 :+1:

**Review by coolchock — APPROVED** [review]
tested in the testing session on ts3, works as expected. Course archive shows old courses, sorting and filtering is possible.

**Review by BBesrour — DISMISSED** [review]
tested on ts3, works as expected

File: src/main/java/de/tum/cit/aet/artemis/core/service/CourseService.java
Comment (edkaya) [reply src/main/java/de/tum/cit/aet/artemis/core/service/CourseService.java:666]: Makes sense to me, I will handle it on a database level 👍

File: src/main/webapp/app/overview/course-archive/course-archive.component.ts
Comment (sarpsahinalp) [inline src/main/webapp/app/overview/course-archive/course-archive.component.ts:16]: According to the client style the Selector should start with jhi?

File: src/main/webapp/app/overview/courses.component.html
Comment (sarpsahinalp) [inline src/main/webapp/app/overview/courses.component.html:65]: We should use the jhiTranslate instead of this way of translation

File: src/main/webapp/app/overview/courses.component.html
Comment (sarpsahinalp) [inline src/main/webapp/app/overview/courses.component.html:66]: Same as above

**Review by sarpsahinalp — COMMENTED** [review]
Some small remarks

**Comment by edkaya** [comment]
> I just tested it on TS3 and noticed that the course images reload quite often. For example, when collapsing a semester in the archive and then opening it again, all course images appear to be loaded again from scratch. Same thing with search, if I remove parts of the search string, any courses that appear will reload their image. Is it possible to cache these images?

We actually use caching strategy for course card icons to load them and cache them in the local storage. For the current local storage, the maxCacheCount is 30, which means that 30 images (or entries) can be stored in localStorage. So in the test servers, there is way more than 30 icons on that page, leading to not to cache all icons. However, I will have a detailed look in a follow up, because this will be the case for course overview as well.

File: src/main/webapp/app/course/manage/course-management.service.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/course/manage/course-management.service.ts:745]: Shouldn't you check semesterB here as well? If both are `WS` you should return 0 here.

File: src/main/webapp/app/overview/course-archive/course-archive.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/overview/course-archive/course-archive.component.html:1]:
```suggestion
@if (courses) {
```
This should do the same

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Code looks good overall.
Please also add some server tests

File: src/main/webapp/app/course/manage/course-management.service.ts
Comment (edkaya) [reply src/main/webapp/app/course/manage/course-management.service.ts:745]: I didn't write that code, just moved that function to the service layer. But I think you are right, it doesn't handle the case when both semesters are the same. I will add this check too 👍

File: src/main/webapp/app/overview/course-archive/course-archive.component.html
Comment (edkaya) [reply src/main/webapp/app/overview/course-archive/course-archive.component.html:1]: This additional check should be necessary, because if the response is null, we declare courses as an empty array. I just moved this check to another part of the code to display the message of "No courses found or smt similar" properly.

**Review by anian03 — COMMENTED** [review]
> We actually use caching strategy for course card icons to load them and cache them in the local storage. For the current local storage, the maxCacheCount is 30, which means that 30 images (or entries) can be stored in localStorage. So in the test servers, there is way more than 30 icons on that page, leading to not to cache all icons. However, I will have a detailed look in a follow up, because this will be the case for course overview as well.

Thanks for the explanation, that makes sense!

I just tested it again and wanted to ask whether the following behavior is intended: By default, the current semester is expanded. When performing a search, all semesters are expanded. However, after clearing the search field, everything remains expanded. Is this intentional? I'm fine with it either way, I just would have expected it to return back to default.

**Review by anian03 — CHANGES_REQUESTED** [review]
Could we hide empty semesters in this list? I think especially for older users with many semesters it can get cluttered quickly if they are preforming a search and just see a bunch of empty semesters (which you can still expand to see even more empty space).



Other than that it works nicely :)

**Comment by edkaya** [comment]
> > We actually use caching strategy for course card icons to load them and cache them in the local storage. For the current local storage, the maxCacheCount is 30, which means that 30 images (or entries) can be stored in localStorage. So in the test servers, there is way more than 30 icons on that page, leading to not to cache all icons. However, I will have a detailed look in a follow up, because this will be the case for course overview as well.
> 
> Thanks for the explanation, that makes sense!
> 
> I just tested it again and wanted to ask whether the following behavior is intended: By default, the current semester is expanded. When performing a search, all semesters are expanded. However, after clearing the search field, everything remains expanded. Is this intentional? I'm fine with it either way, I just would have expected it to return back to default.

Yes, this is an intentional behaviour. If you perform a search and if the a string is matched with a course, its respective semester should be expanded, if there is no match for a semester, that semester should be collapsed. I didn't implement it in a way that it will go back to its old collapse/expand state, but I can have a look

**Comment by edkaya** [comment]
> Could we hide empty semesters in this list? I think especially for older users with many semesters it can get cluttered quickly if they are preforming a search and just see a bunch of empty semesters (which you can still expand to see even more empty space).
> 
> 
> Other than that it works nicely :)

You are right, good catch :) I will remove that additional part if the semester is empty and the user wants to expand it. I am not sure about hiding the semesters because in each search perform, list can be changed too much by hiding and displaying not matched semesters, which may lead to some confusion. I think only collapsing the semesters give users a sense of flexibility, which may allow users to maintain awareness of the overall structure of their course history. Of course, I am still open to discussion 😄

**Comment by anian03** [comment]
> I will remove that additional part if the semester is empty and the user wants to expand it. I am not sure about hiding the semesters because in each search perform, list can be changed too much by hiding and displaying not matched semesters, which may lead to some confusion. I think only collapsing the semesters give users a sense of flexibility, which may allow users to maintain awareness of the overall structure of their course history. Of course, I am still open to discussion 😄

Sounds great! Maybe we can lower the opacity or change to a secondary color to reflect that the semester is empty, that way there isn't as much going on visually 🙂

**Review by janthoXO — CHANGES_REQUESTED** [review]
tested on TS3:
A behaviour that appears with the `Recently accessed courses` : this tab shows the last 3 access courses. when accessing a archived course, it removes one of the recently accessed course, but does not itself get displayed in that tab.

it could be an intended behaviour. But for example if you look at 3 archived courses the tab just disappear.
i think it would be better to also displayed archived courses in the `recently accessed tab` as well.

**Comment by anian03** [comment]
Thanks for the changes! Agreeing with the previous comment, it would be nice if visiting archived courses either doesn't affect "Recently accessed", or if they also somehow appear in that list

**Comment by edkaya** [comment]
> tested on TS3: A behaviour that appears with the `Recently accessed courses` : this tab shows the last 3 access courses. when accessing a archived course, it removes one of the recently accessed course, but does not itself get displayed in that tab.
> 
> it could be an intended behaviour. But for example if you look at 3 archived courses the tab just disappear. i think it would be better to also displayed archived courses in the `recently accessed tab` as well.

Thanks for detecting it 👍 I would suggest we do not display these courses as recently accessed courses, as they are inactive courses and we always want to have active courses in the course overview page. I will fix it by ensuring that recently accessed tab is not affected

**Review by anian03 — DISMISSED** [review]
Works as expected :)

File: src/main/webapp/app/overview/course-archive/course-archive.component.ts
Comment (SimonEntholzer) [inline src/main/webapp/app/overview/course-archive/course-archive.component.ts:20]: Do you really need all those dependencies here? Maybe it's possible to just remove those imports, and only import smaller modules which you actually need

File: src/main/webapp/app/overview/course-archive/course-archive.component.ts
Comment (SimonEntholzer) [inline src/main/webapp/app/overview/course-archive/course-archive.component.ts:44]: Also, here you could use the new features from the new client guidelines. So instead of the constructor, you could use e.g. 
`    private readonly courseService = inject(CourseManagementService);`
You also might be able to get rid of the subscription if you use signals

File: src/main/webapp/app/overview/course-card-header/course-card-header.component.ts
Comment (SimonEntholzer) [inline src/main/webapp/app/overview/course-card-header/course-card-header.component.ts:18]: Here you probably could also remove the imports, and only import what you actually need

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Works during testing, I left a few smaller remarks on the code

File: src/main/webapp/app/overview/course-archive/course-archive.component.ts
Comment (edkaya) [reply src/main/webapp/app/overview/course-archive/course-archive.component.ts:20]: We can import only the required components but the problem is these components need to be also standalone. SearchComponent and its pipe does not standalone and therefore I don't want to change them into standalone in this pr, since it will also affect shared modules.

File: src/main/webapp/app/overview/course-card-header/course-card-header.component.ts
Comment (edkaya) [reply src/main/webapp/app/overview/course-card-header/course-card-header.component.ts:18]: Same here, `ArtemisSharedComponentModule` can be removed but secured image component is not standalone, so I would leave SharedModule in this pr again

**Comment by janthoXO** [comment]
I can sadly not see any archived courses on TS6

**Comment by edkaya** [comment]
> I can sadly not see any archived courses on TS6

Sorry, forgot to explicitly say that you can test with my old courses only on TS3, because i don't have any permission to create old courses on other test servers.

**Review by anian03 — DISMISSED** [review]
Tested on TS3, works as expected :)

**Review by SimonEntholzer — DISMISSED** [review]
Works on Ts3 :+1:

**Review by BBesrour — DISMISSED** [review]

**Review by JohannesStoehr — COMMENTED** [review]
Code looks good overall, left a few comments.
Why are courses without a set semester filtered out? If the student was enrolled and the end date is already over the student might be wondering why one course is missing

File: src/main/java/de/tum/cit/aet/artemis/core/web/CourseResource.java
Comment (edkaya) [reply src/main/java/de/tum/cit/aet/artemis/core/web/CourseResource.java:568]: I just use this for the log print. So it would be nice if it should stay

**Review by SimonEntholzer — APPROVED** [review]
tested again on Ts4, reapprove

**Review by coolchock — APPROVED** [review]
re-approve after merge conflicts were resolved

**Review by BBesrour — APPROVED** [review]
re-approve after fixing coverage

**Review by anian03 — APPROVED** [review]
Re-approve, tested on TS3

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Everything works as expected

Screenshot 2024-10-19 160323
Screenshot 2024-10-19 160401

**Review by rabeatwork — APPROVED** [review]

**Review by zagemello — APPROVED** [review]
image
Tested on TS3, worked as expected

**Review by Cathy0123456789 — APPROVED** [review]
Tested on TS3. Searching, sorting and navigating into the courses work as expected. Nice feature!

**Review by krusche — APPROVED** [review]

**Pull request merged**
